File: plugins/logistics/frontend/api/route-operations.ts
import { API_PREFIX } from "./_shared";
import { apiRequest } from "@systutor/shell/api/client";

export type RouteOperationItem = {
  id: string;
  product_id: string;
  product_name: string;
  quantity: number;
  direction: string;
  created_at: string;
};

export type RouteOperation = {
  id: string;
  session_id: string;
  route_stop_id: string | null;
  context_type: string | null;
  customer_id: string | null;
  customer_name_snapshot: string | null;
  warehouse_id: string | null;
  warehouse_name_snapshot: string | null;
  operation_type: string;
  status: string;
  movement_ids: string[];
  idempotency_key: string;
  location_event_id: string | null;
  location_lat: number | null;
  location_lng: number | null;
  notes: string | null;
  performed_by: string | null;
  performed_at: string | null;
  created_at: string;
  updated_at: string;
  items: RouteOperationItem[];
};

export type RouteOperationCreatePayload = {
  route_stop_id?: string | null;
  operation_type: string;
  notes?: string | null;
  idempotency_key?: string | null;
  items: Array<{
    product_id: string;
    product_name?: string | null;
    quantity: number;
    direction: string;
  }>;
};

export type ExchangeRouteOperationCreatePayload = {
  route_stop_id?: string | null;
  notes?: string | null;
  idempotency_key?: string | null;
  delivered_lines: Array<{
    product_id: string;
    product_name?: string | null;
    quantity: number;
  }>;
  picked_up_lines: Array<{
    product_id: string;
    product_name?: string | null;
    quantity: number;
  }>;
};

export type CurrentComposition = {
  session_id: string;
  composition_version: number | null;
  product_lines: Array<{
    product_id: string;
    product_name: string;
    quantity: number;
    weight_kg: number | null;
    adr_points: number | null;
  }>;
  totals: {
    total_packages: number;
    total_weight_kg: number;
    total_adr_points: number;
  };
};

export type RouteIncident = {
  id: string;
  session_id: string;
  route_stop_id: string | null;
  related_operation_id: string | null;
  type: string;
  status: string;
  corrective_operation_id: string | null;
  notes: string | null;
  created_by: string;
  closed_by: string | null;
  created_at: string;
  closed_at: string | null;
  updated_at: string;
};

export type RouteStopProgress = {
  route_stop_id: string;
  progress_status: string;
  last_operation_at: string | null;
  open_incidents: number;
  completion_percent: number | null;
  outcome_type: string | null;
  driver_note: string | null;
};

export type RouteIncidentCreatePayload = {
  route_stop_id?: string | null;
  related_operation_id?: string | null;
  type: string;
  notes?: string | null;
};

export type RouteIncidentCorrectPayload = {
  route_stop_id?: string | null;
  operation_type: string;
  notes?: string | null;
  idempotency_key?: string | null;
  items: Array<{
    product_id: string;
    product_name?: string | null;
    quantity: number;
    direction: string;
  }>;
};

export type RouteEventConfirmPayload = {
  route_stop_id?: string | null;
  context_type: string;
  customer_id?: string | null;
  warehouse_id?: string | null;
  operation_type: string;
  notes?: string | null;
  idempotency_key: string;
  items: Array<{
    product_id: string;
    product_name?: string | null;
    quantity: number;
    direction: string;
  }>;
  incident_mode: string;
  type?: string | null;
  related_operation_id?: string | null;
  target_incident_id?: string | null;
  incident_notes?: string | null;
};

export function listRouteOperations(sessionId: string) {
  return apiRequest<RouteOperation[]>(`${API_PREFIX}/vehicle-sessions/${sessionId}/route-operations`);
}

export function createRouteOperation(sessionId: string, payload: RouteOperationCreatePayload) {
  return apiRequest<RouteOperation>(`${API_PREFIX}/vehicle-sessions/${sessionId}/route-operations`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function createExchangeRouteOperation(
  sessionId: string,
  payload: ExchangeRouteOperationCreatePayload
) {
  return apiRequest<RouteOperation>(`${API_PREFIX}/vehicle-sessions/${sessionId}/route-operations/exchange`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function confirmRouteOperation(sessionId: string, operationId: string) {
  return apiRequest<RouteOperation>(
    `${API_PREFIX}/vehicle-sessions/${sessionId}/route-operations/${operationId}/confirm`,
    { method: "POST" }
  );
}

export function confirmRouteEvent(sessionId: string, payload: RouteEventConfirmPayload) {
  return apiRequest<RouteOperation>(`${API_PREFIX}/vehicle-sessions/${sessionId}/route-events/confirm`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function getCurrentComposition(sessionId: string) {
  return apiRequest<CurrentComposition>(`${API_PREFIX}/vehicle-sessions/${sessionId}/composition/current`);
}

export function listRouteIncidents(sessionId: string) {
  return apiRequest<RouteIncident[]>(`${API_PREFIX}/vehicle-sessions/${sessionId}/route-incidents`);
}

export function createRouteIncident(sessionId: string, payload: RouteIncidentCreatePayload) {
  return apiRequest<RouteIncident>(`${API_PREFIX}/vehicle-sessions/${sessionId}/route-incidents`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function resolveRouteIncident(sessionId: string, incidentId: string, payload: { notes?: string | null } = {}) {
  return apiRequest<RouteIncident>(
    `${API_PREFIX}/vehicle-sessions/${sessionId}/route-incidents/${incidentId}/resolve`,
    {
      method: "POST",
      body: JSON.stringify(payload),
    }
  );
}

export function correctRouteIncident(sessionId: string, incidentId: string, payload: RouteIncidentCorrectPayload) {
  return apiRequest<RouteIncident>(
    `${API_PREFIX}/vehicle-sessions/${sessionId}/route-incidents/${incidentId}/correct`,
    {
      method: "POST",
      body: JSON.stringify(payload),
    }
  );
}

export function getRouteStopProgress(sessionId: string) {
  return apiRequest<RouteStopProgress[]>(`${API_PREFIX}/vehicle-sessions/${sessionId}/route-stop-progress`);
}

File: plugins/logistics/backend/dto/route_operations.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class RouteOperationItemRequest(BaseModel):
    product_id: str
    product_name: str | None = Field(default=None, max_length=200)
    quantity: float = Field(gt=0)
    direction: str = Field(min_length=1, max_length=10)


class RouteOperationCreateRequest(BaseModel):
    route_stop_id: str | None = None
    operation_type: str = Field(min_length=1, max_length=30)
    notes: str | None = Field(default=None, max_length=500)
    idempotency_key: str | None = Field(default=None, min_length=1, max_length=120)
    items: list[RouteOperationItemRequest] = Field(default_factory=list)


class ExchangeOperationLineRequest(BaseModel):
    product_id: str
    product_name: str | None = Field(default=None, max_length=200)
    quantity: float = Field(gt=0)


class ExchangeRouteOperationCreateRequest(BaseModel):
    route_stop_id: str | None = None
    notes: str | None = Field(default=None, max_length=500)
    idempotency_key: str | None = Field(default=None, min_length=1, max_length=120)
    delivered_lines: list[ExchangeOperationLineRequest] = Field(default_factory=list)
    picked_up_lines: list[ExchangeOperationLineRequest] = Field(default_factory=list)


class RouteOperationItemRead(BaseModel):
    id: str
    product_id: str
    product_name: str
    quantity: float
    direction: str
    created_at: datetime


class RouteOperationRead(BaseModel):
    id: str
    session_id: str
    route_stop_id: str | None = None
    operation_type: str
    status: str
    movement_ids: list[str] = Field(default_factory=list)
    idempotency_key: str
    notes: str | None = None
    performed_by: str | None = None
    performed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    items: list[RouteOperationItemRead] = Field(default_factory=list)


class CompositionLineRead(BaseModel):
    product_id: str
    product_name: str
    quantity: float
    weight_kg: float | None = None
    adr_points: float | None = None


class CompositionTotalsRead(BaseModel):
    total_packages: float
    total_weight_kg: float
    total_adr_points: float


class CurrentCompositionRead(BaseModel):
    session_id: str
    composition_version: int | None = None
    product_lines: list[CompositionLineRead] = Field(default_factory=list)
    totals: CompositionTotalsRead


class RouteIncidentCreateRequest(BaseModel):
    route_stop_id: str | None = None
    related_operation_id: str | None = None
    type: str = Field(min_length=1, max_length=40)
    notes: str | None = Field(default=None, max_length=500)


class RouteIncidentResolveRequest(BaseModel):
    notes: str | None = Field(default=None, max_length=500)


class RouteIncidentCorrectRequest(BaseModel):
    route_stop_id: str | None = None
    operation_type: str = Field(min_length=1, max_length=30)
    notes: str | None = Field(default=None, max_length=500)
    idempotency_key: str | None = Field(default=None, min_length=1, max_length=120)
    items: list[RouteOperationItemRequest] = Field(default_factory=list)


class RouteIncidentRead(BaseModel):
    id: str
    session_id: str
    route_stop_id: str | None = None
    related_operation_id: str | None = None
    type: str
    status: str
    corrective_operation_id: str | None = None
    notes: str | None = None
    created_by: str
    closed_by: str | None = None
    created_at: datetime
    closed_at: datetime | None = None
    updated_at: datetime


class RouteStopProgressRead(BaseModel):
    route_stop_id: str
    progress_status: str
    last_operation_at: datetime | None = None
    open_incidents: int = 0
